File: Extras/compress_newEpisodes.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// ANSI color codes for warnings and errors
const (
	yellowWarning = "\033[33mWarning\033[0m"
	redError      = "\033[31mError\033[0m"
)

// Compiled regular expressions
var (
	patRe  = regexp.MustCompile(`(?i)^(.*),\s*S(\d+)E(\d+),\s*(.*)$`)
	epRe   = regexp.MustCompile(`(?i)\bEpisode\s*(\d{1,4})`)
	ptRe   = regexp.MustCompile(`(?i)\bPart\s*(\d{1,4})`)
	epPtRe = regexp.MustCompile(`(?i)\b(Episode|Part)\s*\d{1,4}`)
	wsRe   = regexp.MustCompile(`\s+`)
)

// ParsedLine represents a parsed episode line
type ParsedLine struct {
	original string
	prefix   string
	season   string
	episode  string
	suffix   string
}

// normalizeQuotedLine removes all quotes from lines that start with '"' and end with '"""'
func normalizeQuotedLine(s string) string {
	trimmed := strings.TrimSpace(s)
	if strings.HasPrefix(s, `"`) && strings.HasSuffix(trimmed, `"""`) {
		return strings.ReplaceAll(s, `"`, "")
	}
	return s
}

// parseLine parses a normalized line into its components
func parseLine(line string) *ParsedLine {
	matches := patRe.FindStringSubmatch(line)
	if matches == nil {
		return nil
	}

	return &ParsedLine{
		original: fmt.Sprintf("%s, S%sE%s, %s", matches[1], matches[2], matches[3], matches[4]),
		prefix:   matches[1],
		season:   matches[2],
		episode:  matches[3],
		suffix:   matches[4],
	}
}

// normalizeLineForComparison removes Episode/Part numbers for comparison
func normalizeLineForComparison(s string) string {
	t := epPtRe.ReplaceAllString(s, "")
	t = wsRe.ReplaceAllString(t, " ")
	t = strings.Trim(t, " :,-")
	return strings.ToLower(t)
}

// isSameArc determines if two lines belong to the same show/season/arc
func isSameArc(a, b *ParsedLine) bool {
	return a.prefix == b.prefix &&
		a.season == b.season &&
		normalizeLineForComparison(a.suffix) == normalizeLineForComparison(b.suffix)
}

// isConsecutiveEpisode checks if all numbers (main E, Episode, Part) are sequential
func isConsecutiveEpisode(a, b *ParsedLine) bool {
	// 1. Check main episode number
	aEp, _ := strconv.Atoi(a.episode)
	bEp, _ := strconv.Atoi(b.episode)
	if bEp != aEp+1 {
		return false
	}

	// 2. Check for 'Episode N' in title
	aEpMatch := epRe.FindStringSubmatch(a.suffix)
	bEpMatch := epRe.FindStringSubmatch(b.suffix)

	if aEpMatch != nil && bEpMatch != nil {
		aNum, _ := strconv.Atoi(aEpMatch[1])
		bNum, _ := strconv.Atoi(bEpMatch[1])
		if bNum != aNum+1 {
			return false
		}
	} else if (aEpMatch == nil) != (bEpMatch == nil) {
		// One has "Episode N" and one doesn't
		return false
	}

	// 3. Check for 'Part N' in title
	aPtMatch := ptRe.FindStringSubmatch(a.suffix)
	bPtMatch := ptRe.FindStringSubmatch(b.suffix)

	if aPtMatch != nil && bPtMatch != nil {
		aNum, _ := strconv.Atoi(aPtMatch[1])
		bNum, _ := strconv.Atoi(bPtMatch[1])
		if bNum != aNum+1 {
			return false
		}
	} else if (aPtMatch == nil) != (bPtMatch == nil) {
		// One has "Part N" and one doesn't
		return false
	}

	return true
}

// getBestWarningNum determines the best number to use when warning about a gap
func getBestWarningNum(p *ParsedLine) string {
	// Check for Part number
	if ptMatch := ptRe.FindStringSubmatch(p.suffix); ptMatch != nil {
		return ptMatch[1]
	}

	// Check for Episode number
	if epMatch := epRe.FindStringSubmatch(p.suffix); epMatch != nil {
		return epMatch[1]
	}

	// Fallback to main episode number
	return p.episode
}

// appendGroup emits a compressed or single line for a group of parsed entries
func appendGroup(group []*ParsedLine, outputLines *[]string) {
	if len(group) == 0 {
		return
	}

	// Single line, nothing to compress
	if len(group) == 1 {
		*outputLines = append(*outputLines, group[0].original)
		return
	}

	first := group[0]
	last := group[len(group)-1]
	e1 := first.episode
	e2 := last.episode
	s1 := first.suffix

	// Extract episode/part numbers
	var e1Text, e2Text, p1Text, p2Text string

	if ep1Match := epRe.FindStringSubmatch(s1); ep1Match != nil {
		e1Text = ep1Match[1]
		if ep2Match := epRe.FindStringSubmatch(last.suffix); ep2Match != nil {
			e2Text = ep2Match[1]
			// Preserve padding
			if strings.HasPrefix(e1Text, "0") && len(e1Text) > 1 {
				num, _ := strconv.Atoi(e2Text)
				e2Text = fmt.Sprintf("%0*d", len(e1Text), num)
			}
		}
	}

	if pt1Match := ptRe.FindStringSubmatch(s1); pt1Match != nil {
		p1Text = pt1Match[1]
		if pt2Match := ptRe.FindStringSubmatch(last.suffix); pt2Match != nil {
			p2Text = pt2Match[1]
			// Preserve padding
			if strings.HasPrefix(p1Text, "0") && len(p1Text) > 1 {
				num, _ := strconv.Atoi(p2Text)
				p2Text = fmt.Sprintf("%0*d", len(p1Text), num)
			}
		}
	}

	// Replace Episode/Part sequences with appropriate ranges
	newSuffix := epPtRe.ReplaceAllStringFunc(s1, func(match string) string {
		parts := epPtRe.FindStringSubmatch(match)
		if len(parts) < 2 {
			return match
		}

		word := parts[1]
		wordLower := strings.ToLower(word)

		if wordLower == "episode" && e1Text != "" && e2Text != "" {
			return fmt.Sprintf("%s %s-%s", word, e1Text, e2Text)
		}
		if wordLower == "part" && p1Text != "" && p2Text != "" {
			return fmt.Sprintf("%s %s-%s", word, p1Text, p2Text)
		}
		return match
	})

	line := fmt.Sprintf("%s, S%sE%s-%s, %s", first.prefix, first.season, e1, e2, newSuffix)
	*outputLines = append(*outputLines, line)
}

// squishLines processes all normalized lines and groups/compresses consecutive entries
func squishLines(lines []string, progName string) []string {
	var outputLines []string
	var group []*ParsedLine

	for _, line := range lines {
		p := parseLine(line)
		if p != nil {
			if len(group) == 0 {
				// Always start a new group if one isn't active
				group = append(group, p)
			} else {
				lastInGroup := group[len(group)-1]
				if isSameArc(lastInGroup, p) {
					// It's the same show/season/arc
					if isConsecutiveEpisode(lastInGroup, p) {
						// It's consecutive, add it to the group
						group = append(group, p)
					} else {
						// GAP DETECTED - same show but not consecutive
						startNum := getBestWarningNum(lastInGroup)
						endNum := getBestWarningNum(p)

						fmt.Fprintf(os.Stderr, "%s: [%s] %s S%s: missing episodes between %s and %s\n",
							progName, yellowWarning, p.prefix, p.season, startNum, endNum)

						// Flush the old group
						appendGroup(group, &outputLines)

						// Start a new group with the current item
						group = []*ParsedLine{p}
					}
				} else {
					// It's a different show/season/arc - normal break
					appendGroup(group, &outputLines)
					group = []*ParsedLine{p}
				}
			}
		} else {
			// Not a parsable line (e.g., a header)
			appendGroup(group, &outputLines)
			group = nil

			// Fix missing space after comma in non-parsable lines
			if strings.Contains(line, ",") && !strings.Contains(line, ", ") {
				line = strings.ReplaceAll(line, ",", ", ")
			}

			outputLines = append(outputLines, line)
		}
	}

	appendGroup(group, &outputLines)
	return outputLines
}

// readLines reads all lines from a reader
func readLines(reader io.Reader) ([]string, error) {
	var lines []string
	scanner := bufio.NewScanner(reader)
	for scanner.Scan() {
		line := normalizeQuotedLine(scanner.Text())
		lines = append(lines, line)
	}
	return lines, scanner.Err()
}

// isTerminal checks if a file descriptor is a terminal
func isTerminal(file *os.File) bool {
	stat, err := file.Stat()
	if err != nil {
		return false
	}
	return (stat.Mode() & os.ModeCharDevice) != 0
}

// printHelp prints the help message
func printHelp(progName string) {
	fmt.Printf(`Reads show episode listings from stdin or a file and compresses consecutive episode sequences.

For example:
  Bannan, S08E01, Episode 01
  Bannan, S08E02, Episode 02
becomes:
  Bannan, S08E01-02, Episode 01-02

Usage:
  %s [input_file]
  %s -h, --help

Arguments:
  input_file    Optional input filename (reads from stdin if omitted)

Options:
  -h, --help    Show this help message and exit

Examples:
  ./newEpisodes.sh | ./%s > newEpisodes.txt
  ./%s episodes.txt > newEpisodes.txt
`, progName, progName, progName, progName)
}

func main() {
	progName := filepath.Base(os.Args[0])

	// Parse arguments
	args := os.Args[1:]

	// Check for help flags
	if len(args) > 0 {
		for _, arg := range args {
			if arg == "-h" || arg == "--help" {
				printHelp(progName)
				os.Exit(0)
			}
		}
	}

	// Check for unknown flags (starting with -)
	for _, arg := range args {
		if strings.HasPrefix(arg, "-") {
			fmt.Fprintf(os.Stderr, "%s: [%s] Unrecognized argument '%s'\n", progName, redError, arg)
			os.Exit(2)
		}
	}

	var reader io.Reader

	if len(args) == 0 {
		// No arguments, check if stdin is a terminal
		if isTerminal(os.Stdin) {
			fmt.Fprintf(os.Stderr, "%s: [%s] No input file or piped input\n", progName, redError)
			fmt.Fprintf(os.Stderr, "Usage: ./%s <file> or use a pipe, e.g.:\n", progName)
			fmt.Fprintf(os.Stderr, "  ./%s episodes.txt\n", progName)
			fmt.Fprintf(os.Stderr, "  ./newEpisodes.sh | ./%s\n", progName)
			os.Exit(2)
		}
		reader = os.Stdin
	} else if len(args) == 1 {
		// One argument - treat as filename
		file, err := os.Open(args[0])
		if err != nil {
			if os.IsNotExist(err) {
				fmt.Fprintf(os.Stderr, "%s: [%s] File '%s' does not exist.\n", progName, redError, args[0])
				os.Exit(1)
			}
			fmt.Fprintf(os.Stderr, "%s: [%s] Error opening file: %v\n", progName, redError, err)
			os.Exit(1)
		}
		defer file.Close() // nolint:errcheck
		reader = file
	} else {
		fmt.Fprintf(os.Stderr, "%s: [%s] Too many arguments\n", progName, redError)
		os.Exit(2)
	}

	// Read and process lines
	lines, err := readLines(reader)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%s: [%s] Error reading input: %v\n", progName, redError, err)
		os.Exit(1)
	}

	squished := squishLines(lines, progName)

	// Write output
	for _, line := range squished {
		fmt.Println(line)
	}
}

File: Extras/compress_newEpisodes.swift
#!/usr/bin/env swift

/*
 compress_newEpisodes.swift is a swift version of compress_newEpisodes.py

 Squish output of newEpisodes.sh to reduce numeric episode sequences
 such that:
     Bannan, S08E01, Episode 01
     Bannan, S08E02, Episode 02
 becomes:
     Bannan, S08E01-02, Episode 01-02
 */

import Foundation

/// ANSI color codes for warnings and errors
let yellowWarning = "\u{001B}[33mWarning\u{001B}[0m"
let redError = "\u{001B}[31mError\u{001B}[0m"

/// Compiled regular expressions
let patRe   = #/(.*),\s*S(\d+)E(\d+),\s*(.*)/#.ignoresCase()
let epRe    = #/\bEpisode\s*(\d{1,4})/#.ignoresCase()
let ptRe    = #/\bPart\s*(\d{1,4})/#.ignoresCase()
let epPtRe  = #/\b(Episode|Part)\s*\d{1,4}/#.ignoresCase()
let wsRe    = #/\s+/#

/// ParsedLine represents a parsed episode line
struct ParsedLine {
    let original: String
    let prefix: String
    let season: String
    let episode: String
    let suffix: String
}

/// Normalize lines that start with '"' and end with '"""' by removing all double quotes
func normalizeQuotedLine(_ s: String) -> String {
    let trimmed = s.trimmingCharacters(in: .whitespaces)
    if s.hasPrefix("\""), trimmed.hasSuffix("\"\"\"") {
        return s.replacingOccurrences(of: "\"", with: "")
    }
    return s
}

/// Parse a normalized line into its components
func parseLine(_ line: String) -> ParsedLine? {
    guard let match = line.wholeMatch(of: patRe) else {
        return nil
    }

    let prefix  = String(match.output.1)
    let season  = String(match.output.2)
    let episode = String(match.output.3)
    let suffix  = String(match.output.4)

    let original = "\(prefix), S\(season)E\(episode), \(suffix)"

    return ParsedLine(original: original, prefix: prefix, season: season, episode: episode, suffix: suffix)
}

/// Remove any Episode/Part numbers before logical grouping comparison
func normalizeLineForComparison(_ s: String) -> String {
    let t = s.replacing(epPtRe, with: "")
    let u = t.replacing(wsRe, with: " ")
    return u.trimmingCharacters(in: CharacterSet(charactersIn: " :,-")).lowercased()
}

/// Determine if two lines belong to the same show/season/arc
func isSameArc(_ a: ParsedLine, _ b: ParsedLine) -> Bool {
    a.prefix == b.prefix &&
        a.season == b.season &&
        normalizeLineForComparison(a.suffix) == normalizeLineForComparison(b.suffix)
}

/// Check if all numbers (main E, Episode, Part) are sequential
func isConsecutiveEpisode(_ a: ParsedLine, _ b: ParsedLine) -> Bool {
    // 1. Check main episode number
    guard let aEp = Int(a.episode), let bEp = Int(b.episode) else {
        return false
    }
    if bEp != aEp + 1 {
        return false
    }

    // 2. Check for 'Episode N' in title
    let aEpMatch = a.suffix.firstMatch(of: epRe)
    let bEpMatch = b.suffix.firstMatch(of: epRe)

    switch (aEpMatch, bEpMatch) {
    case let (aM?, bM?):
        // Both have "Episode N", check if they are consecutive
        guard let aNum = Int(aM.output.1), let bNum = Int(bM.output.1) else { return false }
        if bNum != aNum + 1 { return false }
    case (nil, nil):
        break  // Neither has "Episode N", that's fine
    default:
        return false  // One has "Episode N" and one doesn't
    }

    // 3. Check for 'Part N' in title
    let aPtMatch = a.suffix.firstMatch(of: ptRe)
    let bPtMatch = b.suffix.firstMatch(of: ptRe)

    switch (aPtMatch, bPtMatch) {
    case let (aM?, bM?):
        // Both have "Part N", check if they are consecutive
        guard let aNum = Int(aM.output.1), let bNum = Int(bM.output.1) else { return false }
        if bNum != aNum + 1 { return false }
    case (nil, nil):
        break  // Neither has "Part N", that's fine
    default:
        return false  // One has "Part N" and one doesn't
    }

    return true
}

/// Determine the best number to use when warning about a gap
func getBestWarningNum(_ p: ParsedLine) -> String {
    // Check for Part number
    if let ptMatch = p.suffix.firstMatch(of: ptRe) {
        return String(ptMatch.output.1)
    }

    // Check for Episode number
    if let epMatch = p.suffix.firstMatch(of: epRe) {
        return String(epMatch.output.1)
    }

    // Fallback to main episode number
    return p.episode
}

/// Emit a compressed or single line for a group of parsed entries
func appendGroup(_ group: [ParsedLine], _ outputLines: inout [String]) {
    if group.isEmpty {
        return
    }

    // Single line, nothing to compress
    if group.count == 1 {
        outputLines.append(group[0].original)
        return
    }

    let first = group[0]
    let last  = group[group.count - 1]
    let e1    = first.episode
    let e2    = last.episode
    let s1    = first.suffix

    // Extract episode/part numbers from first and last entries
    var e1Text = "", e2Text = ""
    var p1Text = "", p2Text = ""

    if let ep1 = s1.firstMatch(of: epRe),
       let ep2 = last.suffix.firstMatch(of: epRe) {
        e1Text = String(ep1.output.1)
        e2Text = String(ep2.output.1)

        // Preserve padding
        if e1Text.hasPrefix("0"), e1Text.count > 1, let num = Int(e2Text) {
            e2Text = String(format: "%0\(e1Text.count)d", num)
        }
    }

    if let pt1 = s1.firstMatch(of: ptRe),
       let pt2 = last.suffix.firstMatch(of: ptRe) {
        p1Text = String(pt1.output.1)
        p2Text = String(pt2.output.1)

        // Preserve padding
        if p1Text.hasPrefix("0"), p1Text.count > 1, let num = Int(p2Text) {
            p2Text = String(format: "%0\(p1Text.count)d", num)
        }
    }

    // Replace Episode/Part sequences with appropriate ranges
    // Process matches in reverse to maintain string indices
    var newSuffix = s1
    let matches = s1.matches(of: epPtRe)

    for match in matches.reversed() {
        let word      = String(match.output.1)
        let wordLower = word.lowercased()

        let replacement: String
        if wordLower == "episode", !e1Text.isEmpty, !e2Text.isEmpty {
            replacement = "\(word) \(e1Text)-\(e2Text)"
        } else if wordLower == "part", !p1Text.isEmpty, !p2Text.isEmpty {
            replacement = "\(word) \(p1Text)-\(p2Text)"
        } else {
            continue
        }

        newSuffix.replaceSubrange(match.range, with: replacement)
    }

    let line = "\(first.prefix), S\(first.season)E\(e1)-\(e2), \(newSuffix)"
    outputLines.append(line)
}

/// Process all normalized lines and group/compress consecutive entries
func squishLines(_ lines: [String], _ progName: String) -> [String] {
    var outputLines: [String] = []
    var group: [ParsedLine] = []

    for line in lines {
        if let p = parseLine(line) {
            if group.isEmpty {
                group.append(p)
            } else {
                let lastInGroup = group[group.count - 1]
                if isSameArc(lastInGroup, p) {
                    if isConsecutiveEpisode(lastInGroup, p) {
                        group.append(p)
                    } else {
                        // GAP DETECTED
                        let startNum = getBestWarningNum(lastInGroup)
                        let endNum   = getBestWarningNum(p)

                        fputs("\(progName): [\(yellowWarning)] \(p.prefix) S\(p.season): missing episodes between \(startNum) and \(endNum)\n", stderr)

                        appendGroup(group, &outputLines)
                        group = [p]
                    }
                } else {
                    appendGroup(group, &outputLines)
                    group = [p]
                }
            }
        } else {
            // Not a parsable line
            appendGroup(group, &outputLines)
            group = []

            // Fix missing space after comma in non-parsable lines
            var fixedLine = line
            if line.contains(","), !line.contains(", ") {
                fixedLine = line.replacingOccurrences(of: ",", with: ", ")
            }

            outputLines.append(fixedLine)
        }
    }

    appendGroup(group, &outputLines)
    return outputLines
}

/// Print help message
func printHelp(_ progName: String) {
    print("""
    Reads show episode listings from stdin or a file and compresses consecutive episode sequences.

    For example:
      Bannan, S08E01, Episode 01
      Bannan, S08E02, Episode 02
    becomes:
      Bannan, S08E01-02, Episode 01-02

    Usage:
      \(progName) [input_file]
      \(progName) -h, --help

    Arguments:
      input_file    Optional input filename (reads from stdin if omitted)

    Options:
      -h, --help    Show this help message and exit

    Examples:
      ./newEpisodes.sh | ./\(progName) > newEpisodes.txt
      ./\(progName) episodes.txt > newEpisodes.txt
    """)
}

/// Main entry point
func main() {
    let progName = URL(fileURLWithPath: CommandLine.arguments[0]).lastPathComponent
    let args     = Array(CommandLine.arguments.dropFirst())

    // Check for help flags
    for arg in args {
        if arg == "-h" || arg == "--help" {
            printHelp(progName)
            exit(0)
        }
    }

    // Check for unknown flags
    for arg in args {
        if arg.hasPrefix("-") {
            fputs("\(progName): [\(redError)] Unrecognized argument '\(arg)'\n", stderr)
            exit(2)
        }
    }

    var lines: [String] = []

    if args.isEmpty {
        // Check if stdin is a terminal
        if isatty(STDIN_FILENO) != 0 {
            fputs("\(progName): [\(redError)] No input file or piped input\n", stderr)
            fputs("Usage: ./\(progName) <file> or use a pipe, e.g.:\n", stderr)
            fputs("  ./\(progName) episodes.txt\n", stderr)
            fputs("  ./newEpisodes.sh | ./\(progName)\n", stderr)
            exit(2)
        }

        // Read from stdin
        while let line = readLine(strippingNewline: true) {
            lines.append(normalizeQuotedLine(line))
        }
    } else if args.count == 1 {
        // Read from file
        let filename = args[0]

        guard let content = try? String(contentsOfFile: filename, encoding: .utf8) else {
            fputs("\(progName): [\(redError)] File '\(filename)' does not exist.\n", stderr)
            exit(1)
        }

        lines = content.components(separatedBy: .newlines).map { normalizeQuotedLine($0) }
        // Remove trailing empty line if present
        if lines.last?.isEmpty == true {
            lines.removeLast()
        }
    } else {
        fputs("\(progName): [\(redError)] Too many arguments\n", stderr)
        exit(2)
    }

    let squished = squishLines(lines, progName)

    for line in squished {
        print(line)
    }
}

main()
